File: rust/src/handlers.rs
use std::collections::HashMap;

use axum::{
    Json, Router,
    extract::{Path, Query, State, rejection::JsonRejection},
    http::{StatusCode, header},
    response::{IntoResponse, Response},
    routing::{get, post},
};
use futures::future::join_all;
use serde::Deserialize;
use serde_json::json;

use crate::{
    AppState,
    catalog::ProductError,
    error::AppError,
    store::{CreateUserError, OrderItem},
};

type AppResult<T> = Result<T, AppError>;

pub fn router(state: AppState) -> Router {
    Router::new()
        .route("/health", get(|| async { Json(json!({ "status": "ok" })) }))
        .route("/users", post(create_user))
        .route("/users/{id}", get(get_user))
        .route("/users/{id}/orders", get(list_orders))
        .route("/products/{sku}", get(proxy_product))
        .route("/orders", post(create_order))
        .route("/orders/{id}", get(get_order))
        .with_state(state)
}

fn parse_id(raw: &str) -> AppResult<i64> {
    match raw.parse::<i64>() {
        Ok(id) if id > 0 => Ok(id),
        _ => Err(AppError::BadRequest("invalid id".into())),
    }
}

fn body<T>(payload: Result<Json<T>, JsonRejection>) -> AppResult<T> {
    payload
        .map(|Json(v)| v)
        .map_err(|e| AppError::BadRequest(e.body_text()))
}

#[derive(Deserialize)]
struct CreateUser {
    email: String,
    name: String,
}

async fn create_user(
    State(s): State<AppState>,
    payload: Result<Json<CreateUser>, JsonRejection>,
) -> AppResult<impl IntoResponse> {
    let req = body(payload)?;
    if req.email.len() < 3 || req.email.len() > 254 || !req.email.contains('@') {
        return Err(AppError::BadRequest("invalid email".into()));
    }
    let name_len = req.name.chars().count();
    if !(1..=100).contains(&name_len) {
        return Err(AppError::BadRequest("name must be 1..100 chars".into()));
    }
    match s.store.create_user(&req.email, &req.name).await {
        Ok(u) => Ok((StatusCode::CREATED, Json(u))),
        Err(CreateUserError::Duplicate) => Err(AppError::Conflict("email already exists")),
        Err(CreateUserError::Db(e)) => Err(e.into()),
    }
}

async fn get_user(State(s): State<AppState>, Path(id): Path<String>) -> AppResult<impl IntoResponse> {
    let id = parse_id(&id)?;
    let user = s.store.get_user(id).await?.ok_or(AppError::NotFound("user not found"))?;
    Ok(Json(user))
}

fn query_int(q: &HashMap<String, String>, key: &str, default: i64, min: i64, max: i64) -> AppResult<i64> {
    match q.get(key) {
        None => Ok(default),
        Some(raw) => match raw.parse::<i64>() {
            Ok(v) if (min..=max).contains(&v) => Ok(v),
            _ => Err(AppError::BadRequest(format!("invalid {key}"))),
        },
    }
}

async fn list_orders(
    State(s): State<AppState>,
    Path(id): Path<String>,
    Query(q): Query<HashMap<String, String>>,
) -> AppResult<impl IntoResponse> {
    let id = parse_id(&id)?;
    let limit = query_int(&q, "limit", 20, 1, 100)?;
    let offset = query_int(&q, "offset", 0, 0, i32::MAX as i64)?;
    let (exists, orders) = tokio::join!(s.store.user_exists(id), s.store.list_orders(id, limit, offset));
    if !exists? {
        return Err(AppError::NotFound("user not found"));
    }
    Ok(Json(json!({ "orders": orders?, "limit": limit, "offset": offset })))
}

async fn proxy_product(State(s): State<AppState>, Path(sku): Path<String>) -> Response {
    match s.catalog.get_raw(&sku).await {
        Ok((status, bytes)) => (status, [(header::CONTENT_TYPE, "application/json")], bytes).into_response(),
        Err(e) => AppError::BadGateway(e.to_string()).into_response(),
    }
}

#[derive(Deserialize)]
struct CreateOrderItem {
    sku: String,
    qty: i32,
}

#[derive(Deserialize)]
struct CreateOrder {
    user_id: i64,
    items: Vec<CreateOrderItem>,
}

async fn create_order(
    State(s): State<AppState>,
    payload: Result<Json<CreateOrder>, JsonRejection>,
) -> AppResult<impl IntoResponse> {
    let req = body(payload)?;
    if req.user_id <= 0 || req.items.is_empty() || req.items.len() > 20 {
        return Err(AppError::BadRequest(
            "user_id must be positive and items must have 1..20 entries".into(),
        ));
    }
    if req.items.iter().any(|i| i.sku.is_empty() || !(1..=1000).contains(&i.qty)) {
        return Err(AppError::BadRequest("each item needs a sku and qty in 1..1000".into()));
    }

    // user lookup and all catalog lookups run concurrently
    let (exists, products) = tokio::join!(
        s.store.user_exists(req.user_id),
        join_all(req.items.iter().map(|i| s.catalog.product(&i.sku)))
    );
    if !exists? {
        return Err(AppError::NotFound("user not found"));
    }

    let mut total = 0i64;
    let mut items = Vec::with_capacity(req.items.len());
    let mut currency = String::new();
    for (item, product) in req.items.into_iter().zip(products) {
        let p = match product {
            Ok(p) => p,
            Err(ProductError::NotFound) => return Err(AppError::Unprocessable(format!("unknown sku: {}", item.sku))),
            Err(ProductError::Upstream(e)) => return Err(AppError::BadGateway(e)),
        };
        if i64::from(item.qty) > p.stock {
            return Err(AppError::Unprocessable(format!("insufficient stock: {}", item.sku)));
        }
        total += p.price_cents * i64::from(item.qty);
        currency = p.currency;
        items.push(OrderItem { sku: item.sku, name: p.name, qty: item.qty, unit_price_cents: p.price_cents });
    }

    let order = s.store.create_order(req.user_id, &currency, total, items).await?;
    Ok((StatusCode::CREATED, Json(order)))
}

async fn get_order(State(s): State<AppState>, Path(id): Path<String>) -> AppResult<impl IntoResponse> {
    let id = parse_id(&id)?;
    let order = s.store.get_order(id).await?.ok_or(AppError::NotFound("order not found"))?;
    Ok(Json(order))
}

File: go-sqlite/handlers.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net/http"
	"strconv"
	"strings"
	"sync"
)

type Server struct {
	store   *Store
	catalog *Catalog
}

func (s *Server) Routes() http.Handler {
	mux := http.NewServeMux()
	mux.HandleFunc("GET /health", func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, http.StatusOK, map[string]string{"status": "ok"})
	})
	mux.HandleFunc("POST /users", s.createUser)
	mux.HandleFunc("GET /users/{id}", s.getUser)
	mux.HandleFunc("GET /users/{id}/orders", s.listOrders)
	mux.HandleFunc("GET /products/{sku}", s.proxyProduct)
	mux.HandleFunc("POST /orders", s.createOrder)
	mux.HandleFunc("GET /orders/{id}", s.getOrder)
	return mux
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(v)
}

func writeError(w http.ResponseWriter, status int, msg string) {
	writeJSON(w, status, map[string]string{"error": msg})
}

func internalError(w http.ResponseWriter, err error) {
	log.Printf("internal error: %v", err)
	writeError(w, http.StatusInternalServerError, "internal error")
}

func pathID(w http.ResponseWriter, r *http.Request) (int64, bool) {
	id, err := strconv.ParseInt(r.PathValue("id"), 10, 64)
	if err != nil || id <= 0 {
		writeError(w, http.StatusBadRequest, "invalid id")
		return 0, false
	}
	return id, true
}

func decodeBody(w http.ResponseWriter, r *http.Request, v any) bool {
	if err := json.NewDecoder(io.LimitReader(r.Body, 1<<20)).Decode(v); err != nil {
		writeError(w, http.StatusBadRequest, "invalid json: "+err.Error())
		return false
	}
	return true
}

func (s *Server) createUser(w http.ResponseWriter, r *http.Request) {
	var req struct {
		Email string `json:"email"`
		Name  string `json:"name"`
	}
	if !decodeBody(w, r, &req) {
		return
	}
	if len(req.Email) < 3 || len(req.Email) > 254 || !strings.Contains(req.Email, "@") {
		writeError(w, http.StatusBadRequest, "invalid email")
		return
	}
	if n := len([]rune(req.Name)); n < 1 || n > 100 {
		writeError(w, http.StatusBadRequest, "name must be 1..100 chars")
		return
	}
	u, err := s.store.CreateUser(r.Context(), req.Email, req.Name)
	switch {
	case errors.Is(err, ErrDuplicate):
		writeError(w, http.StatusConflict, "email already exists")
	case err != nil:
		internalError(w, err)
	default:
		writeJSON(w, http.StatusCreated, u)
	}
}

func (s *Server) getUser(w http.ResponseWriter, r *http.Request) {
	id, ok := pathID(w, r)
	if !ok {
		return
	}
	u, err := s.store.GetUser(r.Context(), id)
	switch {
	case errors.Is(err, ErrNotFound):
		writeError(w, http.StatusNotFound, "user not found")
	case err != nil:
		internalError(w, err)
	default:
		writeJSON(w, http.StatusOK, u)
	}
}

func queryInt(r *http.Request, key string, def, min, max int) (int, bool) {
	raw := r.URL.Query().Get(key)
	if raw == "" {
		return def, true
	}
	v, err := strconv.Atoi(raw)
	if err != nil || v < min || v > max {
		return 0, false
	}
	return v, true
}

func (s *Server) listOrders(w http.ResponseWriter, r *http.Request) {
	id, ok := pathID(w, r)
	if !ok {
		return
	}
	limit, ok1 := queryInt(r, "limit", 20, 1, 100)
	offset, ok2 := queryInt(r, "offset", 0, 0, 1<<31-1)
	if !ok1 || !ok2 {
		writeError(w, http.StatusBadRequest, "invalid limit/offset")
		return
	}
	var (
		wg        sync.WaitGroup
		exists    bool
		existsErr error
	)
	wg.Add(1)
	go func() {
		defer wg.Done()
		exists, existsErr = s.store.UserExists(r.Context(), id)
	}()
	orders, err := s.store.ListOrders(r.Context(), id, limit, offset)
	wg.Wait()
	if err == nil {
		err = existsErr
	}
	if err != nil {
		internalError(w, err)
		return
	}
	if !exists {
		writeError(w, http.StatusNotFound, "user not found")
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"orders": orders, "limit": limit, "offset": offset})
}

func (s *Server) proxyProduct(w http.ResponseWriter, r *http.Request) {
	resp, err := s.catalog.Get(r.PathValue("sku"))
	if err != nil {
		writeError(w, http.StatusBadGateway, "upstream unavailable")
		return
	}
	defer resp.Body.Close()
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(resp.StatusCode)
	io.Copy(w, resp.Body)
}

type createOrderReq struct {
	UserID int64 `json:"user_id"`
	Items  []struct {
		SKU string `json:"sku"`
		Qty int32  `json:"qty"`
	} `json:"items"`
}

func (s *Server) createOrder(w http.ResponseWriter, r *http.Request) {
	var req createOrderReq
	if !decodeBody(w, r, &req) {
		return
	}
	if req.UserID <= 0 || len(req.Items) < 1 || len(req.Items) > 20 {
		writeError(w, http.StatusBadRequest, "user_id must be positive and items must have 1..20 entries")
		return
	}
	for _, it := range req.Items {
		if it.SKU == "" || it.Qty < 1 || it.Qty > 1000 {
			writeError(w, http.StatusBadRequest, "each item needs a sku and qty in 1..1000")
			return
		}
	}

	// user lookup and all catalog lookups run concurrently
	var wg sync.WaitGroup
	products := make([]*Product, len(req.Items))
	errs := make([]error, len(req.Items))
	for i, it := range req.Items {
		wg.Add(1)
		go func() {
			defer wg.Done()
			products[i], errs[i] = s.catalog.Product(it.SKU)
		}()
	}
	exists, err := s.store.UserExists(r.Context(), req.UserID)
	wg.Wait()
	if err != nil {
		internalError(w, err)
		return
	}
	if !exists {
		writeError(w, http.StatusNotFound, "user not found")
		return
	}

	items := make([]OrderItem, len(req.Items))
	var total int64
	for i, it := range req.Items {
		if errors.Is(errs[i], ErrProductNotFound) {
			writeError(w, http.StatusUnprocessableEntity, "unknown sku: "+it.SKU)
			return
		}
		if errs[i] != nil {
			log.Printf("catalog error: %v", errs[i])
			writeError(w, http.StatusBadGateway, "upstream error")
			return
		}
		p := products[i]
		if int64(it.Qty) > p.Stock {
			writeError(w, http.StatusUnprocessableEntity, fmt.Sprintf("insufficient stock: %s", it.SKU))
			return
		}
		total += p.PriceCents * int64(it.Qty)
		items[i] = OrderItem{SKU: it.SKU, Name: p.Name, Qty: it.Qty, UnitPriceCents: p.PriceCents}
	}

	o, err := s.store.CreateOrder(r.Context(), req.UserID, products[0].Currency, total, items)
	if err != nil {
		internalError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, o)
}

func (s *Server) getOrder(w http.ResponseWriter, r *http.Request) {
	id, ok := pathID(w, r)
	if !ok {
		return
	}
	o, err := s.store.GetOrder(r.Context(), id)
	switch {
	case errors.Is(err, ErrNotFound):
		writeError(w, http.StatusNotFound, "order not found")
	case err != nil:
		internalError(w, err)
	default:
		writeJSON(w, http.StatusOK, o)
	}
}
